### PR TITLE
Fix a bug in matrix coercion.

### PR DESCRIPTION
**Pull request by wbhart**
One of the issues mentioned in @kirtsar's ticket.

### CODE REVIEW

**Comment by rfourquet** [comment]
I don't find an open ticket from @kirtsar, could you point at it? Also, it would be great to test for the solved bug.

**Comment by wbhart** [comment]
The issue is here:

https://github.com/Nemocas/AbstractAlgebra.jl/issues/367

I'll add some tests to the PR tonight (hopefully).

**Comment by thofma** [comment]
Should we wait for tests or just merge it?

**Comment by rfourquet** [comment]
Not that I want to block this, but the absence of tests makes it difficult to understand what is being fixed by what part of the PR (the mentioned issue seems to report multiple problems). Of course this is due to me being still unfamiliar with the code base, feel free to ignore.
Ideally there would be a test which fails on master and pass with this PR, to be sure that what was intended to be fixed is fixed.

File: src/generic/Matrix.jl [hunk 1]
```diff
@@ -5015,13 +5015,11 @@ function (a::MatSpace{T})(b::Mat{T}) where {T <: RingElement}
...
 
-function (a::MatSpace{T})(b::AbstractArray{T, 2}) where T <: RingElement
+function (a::MatSpace{T})(b::Array{T, 2}) where T <: RingElement
    R = base_ring(a)
    _check_dim(a.nrows, a.ncols, b)
-   for i = 1:a.nrows
-      for j = 1:a.ncols
-         b[i, j] = R(b[i, j])
-      end
+   if (a.nrows != 0 || a.ncols != 0)
```
Comment (rfourquet) [inline src/generic/Matrix.jl]: it should be `if (a.nrows != 0 && a.ncols != 0)`, but even better `if !isempty(b)`

File: src/generic/Matrix.jl [hunk 2]
```diff
@@ -5015,13 +5015,11 @@ function (a::MatSpace{T})(b::Mat{T}) where {T <: RingElement}
    return b
 end
 
-function (a::MatSpace{T})(b::AbstractArray{T, 2}) where T <: RingElement
+function (a::MatSpace{T})(b::Array{T, 2}) where T <: RingElement
```
Comment (rfourquet) [inline src/generic/Matrix.jl]: Why this change?

File: test/generic/Matrix-test.jl [hunk 3]
```diff
@@ -165,6 +176,25 @@ function test_gen_mat_constructors()
    @test x in keys(Dict(x => 1))
    @test !(y in keys(Dict(x => 1)))
 
+   # Test creation from AbstractArray without setindex!
+   A = MyTestMatrix(BigInt(3), 2)
+   S = MatrixSpace(QQ, 2, 2)
+
+   @test isa(S(A), Generic.MatSpaceElem{Rational{BigInt}})
```
Comment (rfourquet) [inline test/generic/Matrix-test.jl]: This seems unrelated to the change, was this not working in the past?

File: src/generic/Matrix.jl (hunk 1 above)
Comment (wbhart) [reply src/generic/Matrix.jl]: Thanks. Fixed.

File: src/generic/Matrix.jl (hunk 2 above)
Comment (wbhart) [reply src/generic/Matrix.jl]: That's all the matrix constructor can accept anyway. And besides, the original issue was that this version was being called, when it certainly should not be.

File: test/generic/Matrix-test.jl (hunk 3 above)
Comment (wbhart) [reply test/generic/Matrix-test.jl]: No, it was not working in the past. This is the issue we are fixing. The original code stopped circulant matrices from SpecialMatrices working, which is what kirtsar reported.

Comment (wbhart) [reply test/generic/Matrix-test.jl]: This test is more general of course and just verifies that any matrix that satisfies the AbstractArray interface, which is beautifully documented on the Julia website, is supported.

File: src/generic/Matrix.jl (hunk 2 above)
Comment (rfourquet) [reply src/generic/Matrix.jl]: > That's all the matrix constructor can accept anyway

I think this change breaks the following:
```julia
julia> m = map(BigInt, sprand(Int8, 2, 3, .9))
2×3 SparseMatrixCSC{BigInt,Int64} with 6 stored entries:
  [1, 1]  =  -11
  [2, 1]  =  -8
  [1, 2]  =  -86
  [2, 2]  =  75
  [1, 3]  =  110
  [2, 3]  =  57

julia> a = MatrixSpace(ZZ, 2, 3)
Matrix Space of 2 rows and 3 columns over Integers

julia> a(m)
[-11  -86  110]
[ -8   75   57]

julia> typeof(ans)
AbstractAlgebra.Generic.MatSpaceElem{BigInt}

julia> @which a(m)
(a::AbstractAlgebra.Generic.MatSpace{T})(b::AbstractArray{T,2}) where T<:Union{RingElem, AbstractFloat, Integer, Rational} in AbstractAlgebra.Generic at /home/atma/.julia/packages/AbstractAlgebra/y3sVP/src/generic/Matrix.jl:5019
```

> the original issue was that this version was being called, when it certainly should not be.

I didn't carefully read the original issue, but a test would be useful to catch that bug, if we want to allow `AbstractArray` again in the future (or are the tests you included already catch that?)

File: test/generic/Matrix-test.jl (hunk 3 above)
Comment (rfourquet) [reply test/generic/Matrix-test.jl]: > This is the issue we are fixing

Ok, but it's not obvious, because this very test passes on AA master without this PR.

Comment (wbhart) [reply test/generic/Matrix-test.jl]: Does it pass if I change it to MatrixSpace(ZZ, 2, 2)?

File: src/generic/Matrix.jl (hunk 2 above)
Comment (wbhart) [reply src/generic/Matrix.jl]: Works for me with the matrix_coerce branch. Are you sure it is broken?

File: test/generic/Matrix-test.jl (hunk 3 above)
Comment (wbhart) [reply test/generic/Matrix-test.jl]: I think it now fails in master but not with this PR.

File: src/generic/Matrix.jl (hunk 2 above)
Comment (rfourquet) [reply src/generic/Matrix.jl]: Nevermind, I guess this method is not called anymore, but there is a more general one right below which catches this example: `(a::MatSpace{T})(b::AbstractArray{S, 2}) where {S <: RingElement, T <: RingElement}`.

Comment (wbhart) [reply src/generic/Matrix.jl]: That is certainly the intention.

File: test/generic/Matrix-test.jl (hunk 3 above)
Comment (rfourquet) [reply test/generic/Matrix-test.jl]: > Does it pass if I change it to MatrixSpace(ZZ, 2, 2)?

no

**Comment by thofma** [comment]
Tests seem to be failing.

**Comment by wbhart** [comment]
@thofma Naturally!

**Comment by thofma** [comment]
Why?

**Comment by wbhart** [comment]
Because I only checked that the later test passed and not the earlier one, after I made the most recent change.
